gen/__init__.py

```diff
@@ -305,20 +305,82 @@ def do_gen_package(config, package_filename):
     log.info("Package filename: %s", package_filename)
 
 
-def extract_files_with_path(start_files, paths):
+def render_late_content(content, late_values):
+    # TODO(branden): Unit tests
+
+    def _next_substring(string_, substring, start=0):
+        idx = string_.find(substring, start)
+        if idx < 0:
+            return None
+        return idx, idx + len(substring)
+
+    def _interpolate_values(string_):
+        if not string_:
+            return
+
+        # Find the next placeholder.
+        placeholder_start = _next_substring(string_, gen.internals.LATE_BIND_PLACEHOLDER_START)
+        if not placeholder_start:
+            # No placeholder found in the string.
+            yield string_
+            return
+
+        # Yield the string preceding the placeholder, if any.
+        if placeholder_start[0] > 0:
+            yield string_[:placeholder_start[0]]
+
+        # Find the end of the placeholder.
+        placeholder_end = _next_substring(string_, gen.internals.LATE_BIND_PLACEHOLDER_END, placeholder_start[1])
+        if not placeholder_end:
+            log.debug("Can't find end of placeholder in string: {}".format(repr(string_)))
+            raise Exception("Bad late config file: Can't find end of placeholder")
+
+        # Extract the name between the start and end and yield its value.
+        name = string_[placeholder_start[1]:placeholder_end[0]]
+        if name not in late_values:
+            log.debug('Found placeholder for unknown value "{}" in string: {}'.format(name, repr(string_)))
+            raise Exception('Bad late config file: Found placeholder for unknown value "{}"'.format(name))
```

Comment:
Just not applicable. This would be a serious error that requires debugging what went wrong between build and cluster bootstrap, and the error may be only incidentally related to config generation.